onnxruntime/contrib_ops/webgpu/moe/qmoe.cc

```diff
@@ -0,0 +1,344 @@
+  const Tensor* hidden_state = context.Input<Tensor>(0);
+  const Tensor* router_logits = context.Input<Tensor>(1);
+  // fc1 is gate_up_proj
+  const Tensor* fc1_experts_weights = context.Input<Tensor>(2);
+  const Tensor* fc1_scales = context.Input<Tensor>(3);
+  const Tensor* fc1_experts_bias_optional = context.Input<Tensor>(4);
+  // fc2 is gate_down_proj
+  const Tensor* fc2_experts_weights = context.Input<Tensor>(5);
+  const Tensor* fc2_scales = context.Input<Tensor>(6);
+  const Tensor* fc2_experts_bias_optional = context.Input<Tensor>(7);
+  const Tensor* fc3_experts_weights_optional = context.Input<Tensor>(8);
+  const Tensor* fc3_scales_optional = context.Input<Tensor>(9);
+  const Tensor* fc3_experts_bias_optional = context.Input<Tensor>(10);
+
+  MoEParameters moe_params;
+
+  ORT_RETURN_IF_ERROR(::onnxruntime::contrib::moe_helper::CheckInputs<Tensor>(
+      moe_params, hidden_state, router_logits,
+      fc1_experts_weights, fc1_experts_bias_optional, fc1_scales,
+      fc2_experts_weights, fc2_experts_bias_optional, fc2_scales,
+      fc3_experts_weights_optional, fc3_experts_bias_optional, fc3_scales_optional,
+      expert_weight_bits_ == 4 ? 2 : 1,
+      activation_type_ == MoEActivationType::SwiGLU, block_size_));
+
+  const auto& input_shape = hidden_state->Shape();
+
+  // SwiGLU validation
+  bool is_swiglu = (activation_type_ == MoEActivationType::SwiGLU);
+  if (is_swiglu && fc3_experts_weights_optional != nullptr) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED,
+                           "SwiGLU activation is not supported with fc3. Gate weights should be concatenated with FC1 weights.");
+  }
+  if (!is_swiglu && fc3_experts_weights_optional != nullptr) {
+    return ORT_MAKE_STATUS(ONNXRUNTIME, NOT_IMPLEMENTED,
+                           "FC3 gating is not yet implemented for non-SwiGLU activations on WebGPU.");
+  }
+
+  const int max_tokens = 256;  // TODO: maybe 512 ?
+  const uint32_t num_experts = static_cast<uint32_t>(moe_params.num_experts);
+  const uint32_t hidden_size = static_cast<uint32_t>(moe_params.hidden_size);
+  const int64_t fc1_output_size = is_swiglu && swiglu_fusion_ > 0 ? 2 * moe_params.inter_size : moe_params.inter_size;
+  const bool is_fp16 = hidden_state->DataType() == DataTypeImpl::GetType<MLFloat16>();
+  const auto dtype = is_fp16 ? DataTypeImpl::GetType<MLFloat16>() : DataTypeImpl::GetType<float>();
+  const auto dtype_uint32 = DataTypeImpl::GetType<uint32_t>();
+
+  const int64_t K_fc1 = moe_params.hidden_size;  // left_shape[left_num_dims - 1]
+  const int64_t N_fc1 = fc1_output_size;         // right_shape[right_num_dims - 1]
+  const int64_t K_fc2 = moe_params.inter_size;   // left_shape[left_num_dims - 1]
+  const int64_t N_fc2 = moe_params.inter_size;   // right_shape[right_num_dims - 1]
+  const int64_t accuracy_level = 4;
+  const int64_t block_size = (block_size_ != 0) ? block_size_ : fc1_experts_weights->Shape()[2];
```

Comment:
Use `const int64_t block_size_fc1 = (block_size_ != 0) ? block_size_ : K_fc1;` for the first ApplyMatMulNBits. And `const int64_t block_size_fc2 = (block_size_ != 0) ? block_size_ : K_fc2;` for the second ApplyMatMulNBits